src/libcore/array.rs

```diff
@@ -62,6 +64,111 @@ unsafe impl<T, A: Unsize<[T]>> FixedSizeArray<T> for A {
     }
 }
 
+/// An iterator that moves out of an array.
+#[derive(Debug)]
+pub struct IntoIter<T, A: FixedSizeArray<T>> {
+    // Invariants: index <= index_back <= array.len()
+    // Only values in array[index..index_back] are alive at any given time.
+    // Values from array[..index] and array[index_back..] are already moved/dropped.
+    array: Option<A>,
+    index: usize,
+    index_back: usize,
+    _marker: PhantomData<T>,
+}
+
+impl<T, A: FixedSizeArray<T>> Drop for IntoIter<T, A> {
+    #[inline]
+    fn drop(&mut self) {
+        // Drop values that are still alive.
+        if let Some(array) = self.array.as_mut() {
+            let slice = array.as_mut_slice();
+            for p in &mut slice[self.index..self.index_back] {
+                unsafe { ptr::drop_in_place(p); }
+            }
+        }
+
+        // Prevent the array as a whole from dropping.
+        unsafe { ptr::write(&mut self.array, None); }
+    }
+}
+
+impl<T, A: FixedSizeArray<T>> Iterator for IntoIter<T, A> {
+    type Item = T;
+
+    #[inline]
+    fn next(&mut self) -> Option<T> {
+        if self.len() > 0 {
+            if let Some(array) = self.array.as_ref() {
+                let slice = array.as_slice();
+                let p = unsafe { slice.get_unchecked(self.index) };
+                self.index += 1;
+                return Some(unsafe { ptr::read(p) })
+            }
+        }
+        None
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.len();
+        (len, Some(len))
+    }
+
+    #[inline]
+    fn count(self) -> usize {
+        self.len()
+    }
+
+    #[inline]
+    fn nth(&mut self, n: usize) -> Option<T> {
+        let len = self.len();
+        if len > 0 {
+            // Drop values prior to the nth.
+            if let Some(array) = self.array.as_mut() {
+                let ndrop = cmp::min(n, len);
+                let slice = array.as_mut_slice();
+                for p in &mut slice[self.index..self.index + ndrop] {
+                    unsafe { ptr::drop_in_place(p); }
```

Comment:
That's a good point!  I think I prefer inside the loop, just to leak as little as possible, but I could go either way.